From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Palladium <ci@tame.gg>
Date: Thu, 4 Jun 2026 12:00:00 -0400
Subject: [PATCH] Palladium startup experience

Banner, onboarding, health snapshot, feature splash, and operator welcome hooks.
---
 .../net/minecraft/server/Main.java                 | 1 +
 .../server/dedicated/DedicatedServer.java          | 1 +
 .../net/minecraft/server/players/PlayerList.java   | 1 +
 3 files changed, 3 insertions(+)

diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 170d2bac2c280921248c86d266add46260c5a76d..8f3a1c2b4d5e6f7a8b9c0d1e2f3a4b5c6d7e8f9a 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -103,6 +103,7 @@ public class Main {
             gg.tame.palladium.config.PalladiumConfig.loadConfig(); // Palladium - Palladium config
 
+            gg.tame.palladium.startup.PalladiumStartup.onConfigLoaded(); // Palladium - startup experience
             CrashReport.preload();
             if (optionSet.has("jfrProfile")) { // CraftBukkit
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index a8970516cd737600ac7fca608ad2d27a083f8dab..b1c2d3e4f5a6b7c8d9e0f1a2b3c4d5e6f7a8b9c0 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -453,6 +453,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             org.purpurmc.purpur.task.BossBarTask.startAll(); // Purpur - Implement TPSBar
             if (org.purpurmc.purpur.PurpurConfig.beeCountPayload) org.purpurmc.purpur.task.BeehiveTask.instance().register(); // Purpur - Give bee counts in beehives to Purpur clients
             if (gg.tame.palladium.config.modules.async.AsyncMobSpawning.enabled) mobSpawnExecutor.start(); // Pufferfish
 
+            gg.tame.palladium.startup.PalladiumStartup.onServerReady(this); // Palladium - startup experience
             this.notificationManager().serverStarted();
             return true;
         }
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 617f5f9742acaf17c9b954009e5086f9302f1735..d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -355,6 +355,7 @@ public abstract class PlayerList {
         org.bukkit.event.player.PlayerJoinEvent playerJoinEvent = new org.bukkit.event.player.PlayerJoinEvent(bukkitPlayer, getJoinMsg(mutableComponent, bukkitPlayer)); // Paper - Adventure // Leaf - Configurable connection message - join message
         this.cserver.getPluginManager().callEvent(playerJoinEvent);
 
+        gg.tame.palladium.startup.PalladiumOperatorWelcome.onPlayerJoin(bukkitPlayer); // Palladium - operator welcome
         if (!player.connection.isAcceptingMessages()) {
             return;
         }
 
         final net.kyori.adventure.text.Component jm = playerJoinEvent.joinMessage();
